src/token/xERC4626.sol

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: MIT
+// Rewards logic inspired by xERC20 (https://github.com/ZeframLou/playpen/blob/main/src/xERC20.sol)
+
+pragma solidity ^0.8.0;
+
+import "solmate/mixins/ERC4626.sol";
+import "solmate/utils/SafeCastLib.sol";
+
+/** 
+ @title  An xERC4626 Single Staking Contract
+ @author Tribe DAO
+ @notice This contract allows users to autocompound rewards denominated in an underlying reward token. 
+         It is fully compatible with ERC4626 allowing for DeFi composability.
+         It maintains balances using internal accounting to prevent instantaneous changes in the exchange rate.
+         NOTE: an exception is at contract creation, when a reward cycle begins before the first deposit. After the first deposit, exchange rate updates smoothly.
+*/
+abstract contract xERC4626 is ERC4626 {
+    using SafeCastLib for *;
+
+    /// @dev thrown when syncing before cycle ends.
+    error SyncError();
+
+    /// @notice the length of a rewards cycle
+    uint32 public immutable rewardsCycleLength;
+
+    /// @notice the delayed start of the current cycle
+    uint32 public lastSync;
+
+    /// @notice the end of the current cycle
+    uint32 public rewardsCycleEnd;
+
+    /// @notice the amount of rewards distributed in a given cycle
+    uint160 public lastRewardAmount;
+
+    uint256 internal storedTotalAssets;
+
+    constructor(uint32 _rewardsCycleLength) {
+        rewardsCycleLength = _rewardsCycleLength;
+        // seed initial rewardsCycleEnd
+        rewardsCycleEnd = uint32(block.timestamp) / rewardsCycleLength * rewardsCycleLength;
+    }
+
+    /// @notice Compute the amount of tokens available to share holders.
+    ///         Increases linearly during a reward distribution period from the sync call, not the cycle start.
+    function totalAssets() public view override returns (uint256) {
+        // cache global vars
+        uint256 storedTotalAssets_ = storedTotalAssets;
+        uint160 lastRewardAmount_ = lastRewardAmount;
+        uint32 rewardsCycleEnd_ = rewardsCycleEnd;
+        uint32 lastSync_ = lastSync;
+        
+        if (block.timestamp >= rewardsCycleEnd_) {
+            // no rewards or rewards fully unlocked
+            // entire reward amount is available
+            return storedTotalAssets_ + lastRewardAmount_;
+        } 
+
+
+        // rewards not fully unlocked
+        // add unlocked rewards to stored total
+        uint256 unlockedRewards = lastRewardAmount_ * (block.timestamp - lastSync_) / (rewardsCycleEnd_ - lastSync_);
+        return storedTotalAssets_ + unlockedRewards;
+    }
+
+    // Update storedTotalAssets on withdraw/redeem
+    function beforeWithdraw(uint256 amount, uint256 shares) internal virtual override {
+        storedTotalAssets -= amount;
```

Comment:
Should the order of these two statements be reversed? You're essentially withdrawing (by subtracting from `storedTotalAssets`) and *then* calling `beforeWithdraw`, which seems backwards.

Reply:
Probably, I think certain edge cases on extending contracts would prefer the ordering you suggest but the generic case for the base class it doesn't matter because storedTotalAssets is just for internal accounting